internal/gitea/team.go

```diff
@@ -0,0 +1,210 @@
+	var orgRepos []repos.RepoDetails
+
+	log.Printf("Connecting to gitea at %s\n", org.URL)
+	client, err := gitea.NewClient(org.URL)
+	if err != nil {
+		return nil, fmt.Errorf("error creating gitea client: %w", err)
+	}
+
+	var repoList []repoMeta
+	if len(org.IncludeRepos) > 0 {
+		repoList, err = specifiedRepos(org, client)
+	} else {
+		repoList, err = publicOrgRepos(org, client)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over repositories, populating release info for each.
+	for _, meta := range repoList {
+		if len(meta.monorepoSources) > 0 {
+			collectedDetails := processFromMonoRepo(
+				client,
+				org.Org,
+				meta.repo,
+				meta.monorepoSources,
+			)
+			for _, details := range collectedDetails {
+				if len(details.Releases) > 0 || len(details.Commits) > 0 {
+					orgRepos = append(orgRepos, details)
+				}
+			}
+		} else {
+			details := processRepo(client, org.Org, meta.repo)
+			if len(details.Releases) > 0 || len(details.Commits) > 0 {
+				orgRepos = append(orgRepos, details)
+			}
+		}
+	}
+
+	return orgRepos, nil
+}
+
+// Process a single (non-mono) repo.
+func processRepo(client *gitea.Client, org string, orgRepo *gitea.Repository) repos.RepoDetails {
+	repo := &Repository{
+		Details: repos.RepoDetails{
+			Name: orgRepo.Name,
+			URL:  orgRepo.HTMLURL,
+		},
+		org:           org,
+		client:        client,
+		defaultBranch: orgRepo.DefaultBranch,
+		folder:        "",
+	}
+
+	log.Printf("processing gitea repo: %s/%s\n", org, repo.Details.Name)
+
+	err := repo.Process()
+	if err != nil {
+		log.Printf("error populating repo '%s' from gitea: %s", repo.Details.Name, err.Error())
+	}
+
+	return repo.Details
+}
+
+// Process multiple 'repositories' from a monorepo.
+func processFromMonoRepo(
+	client *gitea.Client,
+	org string,
+	orgRepo *gitea.Repository,
+	folders []string,
+) []repos.RepoDetails {
+	var collectedDetails []repos.RepoDetails
+
+	// It seems like the gitea client does not have the ability to get only the
+	// top level of the tree, so we have to get the entire tree for the repo.
+	tree, _, err := client.GetTrees(org, orgRepo.Name, orgRepo.DefaultBranch, true)
+	if err != nil {
+		log.Printf("error listing monorepo '%s': %s", orgRepo.Name, err.Error())
+		return collectedDetails
+	}
+
+	for _, folder := range folders {
+		for _, entry := range tree.Entries {
+			parts := strings.Split(entry.Path, "/")
+			if len(parts) < 2 || parts[0] != folder {
+				continue
+			}
+			name := parts[1]
+			if repos.RepoInSlice(collectedDetails, name) {
+				continue
+			}
+
+			repo := &Repository{
+				Details: repos.RepoDetails{
+					Name:     name,
+					URL:      entry.URL,
+					Monorepo: orgRepo.Name,
+				},
+				org:           org,
+				client:        client,
+				defaultBranch: orgRepo.DefaultBranch,
+				folder:        entry.Path,
+			}
+
+			log.Printf("processing gitea sub-repo: %s %s/%s/%s\n",
+				org, orgRepo.Name, folder, repo.Details.Name)
+
+			err := repo.Process()
```

Comment:
I suppose there's a fair bit of duplicated API calls here in a monorepo. Any straight-forward way to cache most of it?

Reply:
The worst part is counting the releases/commits. I'm hesitant to cache that at the moment, because I'd like feedback from @jnsgruk on how he wants that to work in a monorepo first. If it's going to be the Gitea repository release/commit count as whole, then it should definitely get cached. If it's going to be filtered to (e.g.) only commits that contain a file within the `RepoDetails` folder then I think the tree should still be cached but it would be done differently.

I wonder if it would be worth having a function that is essentially `ls {charm-or-snap-folder}`, to separate that out from this (which annoyingly has to run through the individual files). If I wanted a 'only unique' container like a set in Go, would I use a map with nil values?

Reply:
In general information is written out in https://releases.juju.is such that each row of the table is "one package".

I think we should maintain that, or the UI stuff is going to get **really** fiddly - I'm not sure how we'd build a UI that could display all the track/channel info for 30 charms associated with a single repo ;-)

So, I think the best way to approach this (and I'm open for discussion for sure), is the following:

- In the case that the repo is  not declared with `package-dirs: ...`, do as we do today with the other forges
- In the case that we are working with a monorepo, and `package-dirs` are provided, each of those packages should be treated as a separate entity in the list of "repos" (which we may want to name projects, at some point...)

How does that sound?

Reply:
@tonyandrewmeyer To answer your question about an "only unique" container, yes, you typically use a `map[key]bool` or `map[key]struct{}` (the latter is slightly more efficient as the values don't take any space, but usually a bit more awkward to use). There has been talk about a generic `Set` type, but it's not here yet.